crates/autopilot/src/domain/fee/mod.rs

```diff
@@ -55,13 +61,34 @@ impl ProtocolFee {
             }
         }
 
-        let protocol_fees = match &self.policy {
-            policy::Policy::Surplus(variant) => variant.apply(&order, quote),
-            policy::Policy::PriceImprovement(variant) => variant.apply(&order, quote),
-            policy::Policy::Volume(variant) => variant.apply(&order),
-        }
-        .into_iter()
-        .collect_vec();
+        let order_ = boundary::Amounts {
+            sell: order.data.sell_amount,
+            buy: order.data.buy_amount,
+            fee: order.data.fee_amount,
+        };
+        let quote_ = boundary::Amounts {
+            sell: quote.sell_amount,
+            buy: quote.buy_amount,
+            fee: quote.fee,
+        };
+        let protocol_fees = protocol_fees
+            .iter()
+            // TODO: support multiple fee policies
+            .find_map(|fee_policy| {
+                let outside_market_price = boundary::is_order_outside_market_price(&order_, &quote_);
+                match (outside_market_price, fee_policy) {
+                    (true, ProtocolFee::Limit(policy)) => Some(policy),
+                    (false, ProtocolFee::Market(policy)) => Some(policy),
+                    _ => None,
+                }
+            })
+            .and_then(|policy| match policy {
+                policy::Policy::Surplus(variant) => variant.apply(&order),
+                policy::Policy::PriceImprovement(variant) => variant.apply(&order, quote),
+                policy::Policy::Volume(variant) => variant.apply(&order),
+            })
+            .into_iter()
+            .collect_vec();
```

Comment:
nit: this version feels more straight forward to understand, but feel free to ignore:

```
let protocol_fee = match boundary::is_order_outside_market_price(&order_, &quote_) {
            true => protocol_fees.iter().find_map(|fee| match fee {
                ProtocolFee::Limit(policy) => Some(policy),
                _ => None,
            }),
            false => protocol_fees.iter().find_map(|fee| match fee {
                ProtocolFee::Market(policy) => Some(policy),
                _ => None,
            }),
        }
        .and_then(|policy| match policy {
            policy::Policy::Surplus(variant) => variant.apply(&order),
            policy::Policy::PriceImprovement(variant) => variant.apply(&order, quote),
            policy::Policy::Volume(variant) => variant.apply(&order),
        });

        boundary::order::to_domain(order, protocol_fee.into_iter().collect_vec())
```